waiter/src/waiter/instance_tracker.clj

```diff
@@ -0,0 +1,99 @@
+(ns waiter.instance-tracker
+  (:require [clojure.core.async :as async]
+            [clojure.set :as set]
+            [clojure.tools.logging :as log]
+            [metrics.timers :as timers]
+            [waiter.correlation-id :as cid]
+            [waiter.metrics :as metrics]
+            [waiter.util.utils :as utils]))
+
+; Events are being handled by all routers in a cluster for resiliency
+(defprotocol InstanceEventHandler
+
+  (handle-instances-event! [this instances]
+    "handles the instances event")
+
+  (state [this include-flags]
+    "returns the state of the handler:
+    {:last-error-time}"))
+
+(defrecord DefaultInstanceFailureHandler [clock
+                                          handler-state]
+
+  InstanceEventHandler
+
+  (handle-instances-event! [this instances]
+    (let [{:keys [clock handler-state]} this]
+      (log/info "newly failed instances" {:new-failed-instances instances})
+      (reset! handler-state {:last-error-time (clock)})))
+
+  (state [this _]
+    (let [{:keys [handler-state]} this]
+      @handler-state)))
+
+(defn create-instance-failure-event-handler
+  [{:keys [clock]}]
+  (let [handler-state (atom {:last-error-time nil})]
+    (DefaultInstanceFailureHandler. clock handler-state)))
+
+(defn start-instance-tracker
+  "Starts daemon thread that ingests various internal events and routes them to an external source"
+  [clock router-state-chan instance-failure-handler-component]
+  (cid/with-correlation-id
+    "instance-tracker"
+    (let [exit-chan (async/promise-chan)
+          query-chan (async/chan)
+          state-atom (atom {:all-failed-instances #{}
+                            :last-update-time nil})
+          ; create instance tracker component
+          query-state-fn
+          (fn instance-tracker-query-state-fn
+            [include-flags]
+            (let [{:keys [all-failed-instances last-update-time]} @state-atom]
+              (cond-> {:last-update-time last-update-time}
+                      (contains? include-flags "all-failed-instances")
+                      (assoc :all-failed-instances all-failed-instances)
+                      (contains? include-flags "instance-failure-handler")
+                      (assoc :instance-failure-handler (state instance-failure-handler-component [])))))
+          go-chan
+          (async/go
+            (try
+              (loop [{:keys [all-failed-instances] :as current-state} @state-atom]
```

Comment:
`all-failed-instances` can grow very large, there should be a way to clear out old entries which we are no longer care about.

Reply:
all failed instances should reset to the failed instances that are tracked by state-maintainer which caps 10 failed instances per service. So we should not be tracking any more instances than what is already tracked in the state-maintainer. Is this acceptable?

Reply:
The problem with comparing `recent-failed-instances` and the `service-id->failed-instances` provided with by the state-maintainer is that the tracker might think some failing instances are new but they are really an hour older

Reply:
I missed that we are resetting the value to instances from `service-id->failed-instances`. Failed instances cannot grow indefinitely large.